### PR TITLE
[Fix] keras.ops.image.extract_patches returns incorrect output shape for channels_first

### PR DESCRIPTION
**Pull request by ChiragSW**
Fixes issue: #22531

### CODE REVIEW

**Comment by ChiragSW** [comment]
Please review, the issue has been resolved

**Comment by keerthanakadiri** [comment]
Hi @ChiragSW, Can you please provide more information and how this PR would solve the issue.

**Comment by ChiragSW** [comment]
## Root cause

Root cause of the issue is that ```keras.ops.image.extract_patches``` produced inconsistent and incorrect output shapes when using the ```data_format="channels_first"``` configuration, leading to a mismatch between eager execution and symbolic static shapes.

## Fix 

Updated the logic of the ```extract_patches``` op to correctly handle the axis ordering and spatial dimension calculations specifically for the NCHW (channels_first) format, so that the inferred static shape matches the actual runtime output tensor across all execution modes.

**Comment by ChiragSW** [comment]
@keerthanakadiri any updates needed?

File: keras/src/ops/image.py [hunk 1]
```diff
@@ -845,14 +892,20 @@ def _extract_patches_3d(
...
+        volumes = backend.numpy.transpose(volumes, (0, 2, 3, 4, 1))
+        conv_data_format = "channels_last"
     patches = backend.nn.conv(
         inputs=volumes,
         kernel=kernel,
         strides=strides,
         padding=padding,
-        data_format=data_format,
+        data_format=conv_data_format,
         dilation_rate=dilation_rate,
```
Comment (hertschuh) [inline keras/src/ops/image.py]: What this points to is that the tensorflow implementation of `nn.conv` has an issue with the `data_format` (although at first sight the code looks fine). So this is where the bug fix should go, not here.

File: keras/src/ops/image.py [hunk 2]
```diff
@@ -831,11 +860,29 @@ def _extract_patches_3d(
...
+        volumes_shape = [1] + volumes_shape
+    compute_conv_output_shape(
+        volumes_shape,
+        filters=1,
+        kernel_size=(patch_d, patch_h, patch_w),
+        strides=strides,
+        padding=padding,
+        data_format=data_format,
+        dilation_rate=dilation_rate,
+    )
```
Comment (hertschuh) [inline keras/src/ops/image.py]: You're not using the result of `compute_conv_output_shape`, just remove this.

**Comment by ChiragSW** [comment]
The suggested changes are done, please review @hertschuh

File: keras/src/backend/tensorflow/nn.py [hunk 3]
```diff
@@ -791,6 +791,47 @@ def conv(
...
+            data_format = "channels_last"
+            result = conv(
+                inputs,
+                kernel,
+                strides=strides,
+                padding=padding,
+                data_format=data_format,
+                dilation_rate=dilation_rate,
+            )
+            return tf.transpose(result, (0, 4, 1, 2, 3))
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/nn.py:833]: `data_format` is handled below.

What is this intended to fix?

File: keras/src/ops/image.py [hunk 4]
```diff
@@ -795,14 +822,30 @@ def _extract_patches_2d(
...
+    except (ValueError, RuntimeError) as e:
+        message = str(e)
+        message_lower = message.lower()
+        if (
+            "empty output" in message_lower
+            or "computed output size would be zero or negative" in message_lower
+            or "kernel after dilation has size" in message_lower
+            or "kernel size can't be greater than actual input size"
+            in message_lower
+            or "calculated padded input size per channel" in message_lower
```
Comment (hertschuh) [inline keras/src/ops/image.py:843]: This is extremely brittle. I think I'd rather let the exception go through even if the message is a bit unclear in the context of `extract_patches`.

File: keras/src/ops/image.py [hunk 5]
```diff
@@ -784,9 +784,36 @@ def _extract_patches_2d(
...
+        try:
+            compute_conv_output_shape(
+                images_shape,
+                out_dim,
+                (patch_h, patch_w),
+                strides=strides,
+                padding=padding,
+                data_format=data_format,
+                dilation_rate=dilation_rate,
+            )
```
Comment (hertschuh) [inline keras/src/ops/image.py:814]: Please add a comment saying this is just for validating the parameters.

File: keras/src/ops/image.py [hunk 6]
```diff
@@ -784,9 +784,36 @@ def _extract_patches_2d(
...
+        except TypeError:
+            pass
```
Comment (hertschuh) [inline keras/src/ops/image.py:816]: Why ignore `TypeError`? When does it happen?

**Comment by ChiragSW** [comment]
Actually the bug was fixed. This PR is good to close. Sorry for the wrong overhead
